x-pack/legacy/plugins/infra/public/containers/logs/log_filter/index.ts

```diff
@@ -4,23 +4,114 @@
+import { esKuery } from '../../../../../../../../src/plugins/data/public';
 
-export const useLogFilterState = () => {
-  const { local: state } = useContext(ReduxStateContext);
-  const filterQuery = logFilterReduxSelectors.selectLogFilterQueryAsJson(state);
-  return { filterQuery };
-};
+enum Action {
+  SetDraftFilterQuery,
+  ApplyFilterQuery,
+}
+
+interface SetDraftQueryAction {
+  type: Action.SetDraftFilterQuery;
+  payload: FilterQuery | null;
+}
+
+interface ApplyFilterQueryAction {
+  type: Action.ApplyFilterQuery;
+  payload: SerializedFilterQuery;
+}
+
+type ActionObj = SetDraftQueryAction | ApplyFilterQueryAction;
 
-export interface LogFilterStateParams {
-  filterQuery: string | null;
+export interface KueryFilterQuery {
+  kind: 'kuery';
+  expression: string;
 }
 
-export const logFilterInitialState = {
+export type FilterQuery = KueryFilterQuery;
+
+export interface SerializedFilterQuery {
+  query: FilterQuery;
+  serializedQuery: string;
+}
+
+interface LogFilterBaseStateParams {
+  filterQuery: SerializedFilterQuery | null;
+  filterQueryDraft: KueryFilterQuery | null;
+}
+
+export const logFilterInitialState: LogFilterBaseStateParams = {
   filterQuery: null,
+  filterQueryDraft: null,
+};
+
+interface LogFilterDerivatives {
+  isFilterQueryDraftValid: boolean;
+  filterQueryAsJson: SerializedFilterQuery['serializedQuery'] | null;
+  filterQuery: FilterQuery | null;
+}
+
+export type LogFilterStateParams = Omit<LogFilterBaseStateParams, 'filterQuery'> &
+  LogFilterDerivatives;
+
+export interface LogFilterCallbacks {
+  setLogFilterQueryDraft: (payload: FilterQuery) => void;
+  applyLogFilterQuery: (payload: SerializedFilterQuery) => void;
+}
+
+const getDerivatives = (state: LogFilterBaseStateParams) => ({
+  get isFilterQueryDraftValid() {
+    const { filterQueryDraft } = state;
+    if (filterQueryDraft && filterQueryDraft.kind === 'kuery') {
+      try {
+        esKuery.fromKueryExpression(filterQueryDraft.expression);
+      } catch (err) {
+        return false;
+      }
+    }
+
+    return true;
+  },
+  get filterQueryAsJson() {
+    const { filterQuery } = state;
+    return filterQuery ? filterQuery.serializedQuery : null;
+  },
+  get filterQuery() {
+    const { filterQuery } = state;
+    return filterQuery ? filterQuery.query : null;
+  },
+});
+
+export const useLogFilterState: () => [LogFilterStateParams, LogFilterCallbacks] = () => {
+  const [state, dispatch] = useReducer(logFilterStateReducer, logFilterInitialState);
+
+  const callbacks = useMemo(
+    () => ({
+      setLogFilterQueryDraft: (payload: FilterQuery) =>
+        dispatch({ type: Action.SetDraftFilterQuery, payload }),
+
+      applyLogFilterQuery: (payload: SerializedFilterQuery) =>
+        dispatch({ type: Action.ApplyFilterQuery, payload }),
+    }),
+    []
+  );
+
+  const derivatives = useMemo(() => getDerivatives(state), [state]);
+
+  return [{ ...state, ...derivatives }, callbacks];
```

Comment:
[nit] is there a reason to use an array here?

Reply:
Consistency with https://github.com/elastic/kibana/blob/master/x-pack/legacy/plugins/infra/public/containers/logs/log_entries/index.ts#L214-L222 and with the `[value, updaterFunction]` pattern of hooks in general. Do you have another suggestion for what pattern to use?

Reply:
I like that pattern for exactly two return values of the `[value, setter]`. For more complex return values I'd go with an object, since that forces us to name the constituents descriptively.